### DISCUSSION TITLE
Change HTML errors to JSON

### DISCUSSION THREAD

**Discussion by a1exk-dev** [symfony/symfony#42441] [Q&A] [answered]
Hi guys! A quick question:
I have a twig installed in order for NelmioApiDocBundle to render its UI.
I have a simple Rest controller (Symfony 5.3 + PHP8):

```php
class LanguagesDictionaryController extends AbstractController
{
    #[Route('/api/dictionaries/language', name: 'dictionaries/get_languages', methods: 'GET', )]
    public function getLanguagesDictionary(): JsonResponse
    {
        return $this->json('...some data');
    }
}
```

But whenever for example I try to POST, PUT, or else the error is in HTML format instead of JSON?
Even setting `format: json` changes nothing.
I don't need HTML errors, the app is RestAPI only.

How could I possibly archive that?

**Comment by faizanakram99** [comment, marked as answer]
You need to install serializer in order for Symfony's json error handler to kick in.

See https://symfony.com/doc/current/controller/error_pages.html#overriding-error-output-for-non-html-formats 

Other alternative would be to create an exception handler which would set response to json.

As an example.


```php
class JSONExceptionListener implements EventSubscriberInterface
{
    /**
     * @var string
     */
    private $environment;

    public function __construct(string $environment)
    {
        $this->environment = $environment;
    }

    public static function getSubscribedEvents()
    {
        return [
            KernelEvents::EXCEPTION => ['onKernelException', 200],
        ];
    }

    public function onKernelException(ExceptionEvent $event)
    {
        $exception = $event->getThrowable();
        
        $content = [
            'code' => $exception instanceof HttpExceptionInterface ? $exception->getStatusCode() : 500,
            'message' => $exception->getMessage(),
            'trace' => \in_array($this->environment, ['dev', 'test'], true)
                ? $exception->getTrace()
                : [],
        ];

        $event->setResponse(
            new JsonResponse($content, $content['code'])
        );
    }
}

```

**Reply by dbrumann** [reply]
> You need to install serializer in order for Symfony's json error handler to kick in.

You will also need to set the `Accept`-header on your request with an appropriate value, e.g. `Accept: application/json` for Symfony to know that it should render the output as JSON instead of HTML.

**Comment by javiereguiluz** [comment]
@alexanderkatkov if any of the provided answers solved your issue, please click on their "Mark as answer" icon. Thanks!

**Reply by a1exk-dev** [reply]
@faizanakram99, @dbrumann thanks for the answers will try one of the options today or tomorrow.
@javiereguiluz I will check the best option and write about the variant I went in my case, so let's keep it open for 1 more day.
